fluent-bundle/src/resolve.rs

```diff
@@ -13,271 +13,294 @@ use std::hash::{Hash, Hasher};
 
 use super::bundle::FluentBundle;
 use super::entry::GetEntry;
+use super::types::DisplayableNode;
 use super::types::FluentValue;
 use fluent_syntax::ast;
 use fluent_syntax::unicode::unescape_unicode;
 
-#[derive(Debug, PartialEq)]
+#[derive(Debug, PartialEq, Clone)]
 pub enum ResolverError {
-    None,
+    Reference(String),
+    MissingDefault,
+    Argument(String),
     Value,
     Cyclic,
 }
 
 /// State for a single `ResolveValue::to_value` call.
-pub struct Env<'env> {
+pub struct Scope<'scope> {
     /// The current `FluentBundle` instance.
-    pub bundle: &'env FluentBundle<'env>,
+    pub bundle: &'scope FluentBundle<'scope>,
     /// The current arguments passed by the developer.
-    pub args: Option<&'env HashMap<&'env str, FluentValue>>,
+    pub args: Option<&'scope HashMap<&'scope str, FluentValue<'scope>>>,
+    /// Local args
+    pub local_args: Option<HashMap<&'scope str, FluentValue<'scope>>>,
     /// Tracks hashes to prevent infinite recursion.
     pub travelled: RefCell<Vec<u64>>,
+    /// Track errors accumulated during resolving.
+    pub errors: Vec<ResolverError>,
 }
 
-impl<'env> Env<'env> {
+impl<'scope> Scope<'scope> {
     pub fn new(
-        bundle: &'env FluentBundle,
-        args: Option<&'env HashMap<&'env str, FluentValue>>,
+        bundle: &'scope FluentBundle<'scope>,
+        args: Option<&'scope HashMap<&str, FluentValue>>,
     ) -> Self {
-        Env {
+        Scope {
             bundle,
             args,
+            local_args: None,
             travelled: RefCell::new(Vec::new()),
+            errors: vec![],
         }
     }
 
-    fn track<F>(&self, identifier: &str, action: F) -> Result<FluentValue, ResolverError>
+    pub fn track<F>(&mut self, entry: DisplayableNode<'scope>, mut action: F) -> FluentValue<'scope>
     where
-        F: FnMut() -> Result<FluentValue, ResolverError>,
+        F: FnMut(&mut Scope<'scope>) -> FluentValue<'scope>,
     {
         let mut hasher = DefaultHasher::new();
-        identifier.hash(&mut hasher);
+        entry.id.hash(&mut hasher);
+        if let Some(attr) = entry.attribute {
+            attr.hash(&mut hasher);
+        }
         let hash = hasher.finish();
 
         if self.travelled.borrow().contains(&hash) {
-            Err(ResolverError::Cyclic)
+            self.errors.push(ResolverError::Cyclic);
+            FluentValue::Error(entry)
         } else {
             self.travelled.borrow_mut().push(hash);
-            self.scope(action)
+            let result = action(self);
+            self.travelled.borrow_mut().pop();
```

Comment:
Could this instead be a HashSet with us removing the element when done? That seems faster on the `.contains()` call, plus we don't have to manually hash anything.

The only hazard with this is that cycles will reuse the same entry for the same element, however you forbid cycles above so it's fine.

The `Vec` isn't necessary here since we already have the call stack for keeping track of what got pushed last. A `Vec` would be useful if we decided to abandon recursion in favor of iteration, but if we're not doing that we should just use a HashSet and take the win on the `.contains()` call

Reply:
I tried to do that, and got a significant perf regression. Separated out to https://github.com/projectfluent/fluent-rs/issues/100